File: 백준/Silver/11725. 트리의 부모 찾기/트리의 부모 찾기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {

    static int[] parent;
    static List<Integer>[] graph;

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        int N = Integer.parseInt(br.readLine());
        graph = new ArrayList[N + 1];

        for (int i = 1; i <= N; i++) {
            graph[i] = new ArrayList<>();
        }

        for (int i = 0; i < N - 1; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            graph[a].add(b);
            graph[b].add(a);
        }

        parent = new int[N + 1];
        parent[1] = -1;

        findParent(1);
        for (int i = 2; i <= N; i++) {
            sb.append(parent[i]).append("\n");
        }
        System.out.println(sb);


    }

    private static void findParent(int i) {
        for (int node : graph[i]) {
            if (parent[node] == 0) {
                parent[node] = i;
                findParent(node);
            }
        }
    }


}

File: 백준/Silver/11725. 트리의 부모 찾기/트리의 부모 찾기.py
import sys
input = sys.stdin.readline
sys.setrecursionlimit(10**6)

N = int(input())  # 노드 수

graph = [[] for _ in range(N + 1)]
parent = [0] * (N + 1)
parent[1] = -1

for _ in range(N - 1):
    a, b = map(int, input().split())
    graph[a].append(b)
    graph[b].append(a)


def find_parent(node):
    for next in graph[node]:
        if parent[next] == 0:
            parent[next] = node
            find_parent(next)


find_parent(1)


# list(map(print, parent[2:N+1]))
print("\n".join(map(str, parent[2:N+1])))
